### PR TITLE
feat(CLDX-79): add sign-mac-binaries step to sign-binaries task

### PR DESCRIPTION
**Pull request by pkhander**
This commit:

- adds sign-mac-binaries step to the **sign-binaries** task.
- adds push-unsigned-using-oras step to the  **sign-binaries** task.
- adds logic to handle checksum files to the **publish-to-cgw** task.

### CODE REVIEW

File: tasks/sign-binaries/sign-binaries.yaml [hunk 1]
```diff
@@ -42,36 +45,203 @@ spec:
...
+      secret:
+        secretName: $(params.macSSHKey)
   workspaces:
     - name: data
       description: Workspace to save the results to
   results:
+    - name: unsignedMacDigest
+      type: string
+      description: |
+        Digest used by Mac signing host to pull unsignged content via ORAS
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: typo unsignged

File: tasks/sign-binaries/sign-binaries.yaml [hunk 2]
```diff
@@ -42,36 +45,203 @@ spec:
...
+    - name: signedMacDigest
+      type: string
+      description: |
+        Digest used to pull signed Mac content back to pipeline via ORAS
     - name: unsignedWindowsDigest
       type: string
       description: |
-        Digest used by signing host to pull unsignged content via ORAS
+        Digest used by Windows signing host to pull unsignged content via ORAS
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: typo unsignged

File: tasks/sign-binaries/sign-binaries.yaml [hunk 3]
```diff
@@ -42,36 +45,203 @@ spec:
...
+      script: |
+        #!/usr/bin/env bash
+        set -eux
 
-    # - name: sign-mac-binaries
-    #   image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
-    #   script: |
-    #     #!/usr/bin/env bash
-    #     # TODO CLDX-79
+        mkdir -p /root/.ssh
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml:235]: Have you tested this? I don't think the normal user has perms to write to `/root` in a tekton taskrun pod in our deployment

File: tasks/sign-binaries/sign-binaries.yaml [hunk 4]
```diff
@@ -42,36 +45,203 @@ spec:
...
+        chmod 700 /root/.ssh
+        cp "/etc/secrets/mac_id_rsa" /root/.ssh/id_rsa
+        cp "/etc/secrets/mac_fingerprint" /root/.ssh/known_hosts
+        chmod 600 root/.ssh/known_hosts /root/.ssh/id_rsa
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: missing `/` before the known hosts file

**Review by johnbieren — COMMENTED** [review]
This stuff is pretty far out of my domain. Is there anyone from your team who knows it better who can also review it?

**Comment by pkhander** [comment]
> This stuff is pretty far out of my domain. Is there anyone from your team who knows it better who can also review it?

Mac-signing is currently done in ETERA (Ansible) in the traditional pipeline. This is mostly porting the logic here in bash and tekton. @swickersh would a good one to nominate to help review here :) 

cc - @deekej

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml:235]: I am working on testing this with the pipeline currently. Will update here once i am done, ty!

File: tasks/sign-binaries/sign-binaries.yaml [hunk 5]
```diff
@@ -4,14 +4,17 @@ kind: Task
 metadata:
   name: sign-binaries
   labels:
-    app.kubernetes.io/version: "1.0.0"
+    app.kubernetes.io/version: "1.2.0"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Why increment by 2? Should be 1.1.0?

File: tasks/sign-binaries/sign-binaries.yaml [hunk 6]
```diff
@@ -42,36 +45,203 @@ spec:
...
       description: |
-        Digest used to pull signed content back to pipeline via ORAS
+        Digest used to pull signed Windows content back to pipeline via ORAS
   steps:
     - name: push-unsigned-using-oras
       image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
       script: |
-        #!/usr/bin/env bash
+        #!/bin/bash
+        set -x
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Can we use `set -eux`? That way if something fails along the way, the task will stop there, which is typically what we want.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 7]
```diff
@@ -42,36 +45,203 @@ spec:
...
+
+        CONTENT_DIR=$(workspaces.data.path)/$(params.pipelineRunUid)/$(params.contentDir)
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: I would suggest using just `CONTENT_DIR=$(workspaces.data.path)/$(params.contentDir)` and have the parameter point to the full relative path.

But I wonder where this will come from. There will be another task that will fetch the binaries from somewhere? From where? Also, currently the workspace uses a PVC that is 1 GB in size. Will that be enough?

File: tasks/sign-binaries/sign-binaries.yaml [hunk 8]
```diff
@@ -42,36 +45,203 @@ spec:
...
       type: string
       description: |
-        Digest used to pull signed content back to pipeline via ORAS
+        Digest used to pull signed Windows content back to pipeline via ORAS
   steps:
     - name: push-unsigned-using-oras
       image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
       script: |
-        #!/usr/bin/env bash
+        #!/bin/bash
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Why this change? It doesn't really make a difference, but we use `#!/usr/bin/env bash` everywhere, so it's nice to stick with that.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 9]
```diff
@@ -42,36 +45,203 @@ spec:
...
-    #   script: |
-    #     #!/usr/bin/env bash
-    #     # TODO CLDX-79
+        mkdir -p /root/.ssh
+        chmod 700 /root/.ssh
+        cp "/etc/secrets/mac_id_rsa" /root/.ssh/id_rsa
+        cp "/etc/secrets/mac_fingerprint" /root/.ssh/known_hosts
+        chmod 600 root/.ssh/known_hosts /root/.ssh/id_rsa
+
+        #TODO: cp SSH key to the container
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: To which container? You already copy it to `/root/.ssh/id_rsa` above.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 10]
```diff
@@ -42,36 +45,203 @@ spec:
...
+        # shell check complains about the variable (unsigned_digest) not being used but it is used in the script
+        # shellcheck disable=SC2034
+        unsigned_digest=$(cat "$(results.unsignedMacDigest.path)")
+        mac_signing_script="/tmp/mac_signing_script.sh"
+        set +x
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: This is not needed (and the matching `set -x` after the EOF line). The contents of the here-document are not printed either way.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 11]
```diff
@@ -42,36 +45,203 @@ spec:
...
+
+        PUSH_OUTPUT=$(oras push "$(params.quayURL)/signed:$(params.pipelineRunUid)" "$CONTENT_DIR")
+        SIGNED_DIGEST=$(echo "$PUSH_OUTPUT" | grep 'Digest:' | awk '{print $2}')
+        echo "$SIGNED_DIGEST" >> "$DIGEST_FILE"
+        echo "Process completed successfully."
+        EOF
+        set -x
+        # Copy the script to the Mac host
+        scp -i "/etc/secrets/id_rsa" "$mac_signing_script" \
+        "${MAC_USER}@${MAC_HOST}:/tmp/mac_signing_script.sh"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Indent for better readability please.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 12]
```diff
@@ -42,36 +45,203 @@ spec:
...
+
+        # Execute the script on the Mac host
+        ssh -i "/etc/secrets/id_rsa" "${MAC_USER}@${MAC_HOST}" \
+        "bash /tmp/mac_signing_script.sh"
+
+        # Copy the signed digest back to the pipeline
+        scp -i "/etc/secrets/id_rsa" "${MAC_USER}@${MAC_HOST}:/tmp/$(params.pipelineRunUid)/push_digest.txt" \
+        "$(results.signedMacDigest.path)"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Indent for better readability please.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 13]
```diff
@@ -42,36 +45,203 @@ spec:
...
+        echo "Process completed successfully."
+        EOF
+        set -x
+        # Copy the script to the Mac host
+        scp -i "/etc/secrets/id_rsa" "$mac_signing_script" \
+        "${MAC_USER}@${MAC_HOST}:/tmp/mac_signing_script.sh"
+
+        # Execute the script on the Mac host
+        ssh -i "/etc/secrets/id_rsa" "${MAC_USER}@${MAC_HOST}" \
+        "bash /tmp/mac_signing_script.sh"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Indent for better readability please.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 14]
```diff
@@ -42,36 +45,203 @@ spec:
...
+            --password "$APP_SPECIFIC_PASSWORD"
+
+        PUSH_OUTPUT=$(oras push "$(params.quayURL)/signed:$(params.pipelineRunUid)" "$CONTENT_DIR")
+        SIGNED_DIGEST=$(echo "$PUSH_OUTPUT" | grep 'Digest:' | awk '{print $2}')
+        echo "$SIGNED_DIGEST" >> "$DIGEST_FILE"
+        echo "Process completed successfully."
+        EOF
+        set -x
+        # Copy the script to the Mac host
+        scp -i "/etc/secrets/id_rsa" "$mac_signing_script" \
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Why do you need the `-i`? You copied the private key to `/root/.ssh/id_rsa` so that should work, right? Also, it will probably not work from this location since it won't have the required permissions. This applies to the others as well.

Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: When I was running tests for the Windows tasks, `whoami` from the container runs returned `root` but the home directory for root was `/home/tekton`

During testing when I put the id_rsa into `~/.ssh` or `/home/tekton/.ssh` it wouldn't find it by default. (and it should have per the ssh_config)

But when I put it at `/root/.ssh/` it worked fine.

Maybe the containers run with a different user in prod? My tests were all just ran locally and when I exec into a running container and look at /etc/passwd I didn't see any other users..

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: OK, yeah, I guess whatever works is fine. But it doesn't make sense to copy the key to another place, but then not use that.

Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Yep I agree. This should be `scp -i "/root/.ssh/id_rsa" `

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml:235]: this was tested locally now and it worked. 

by locally I mean, we created a dummy workspace, applied the task and a dummy pipeline, ran the pipeline and it gave us the expected results. 

```
  results:
  - name: unsignedMacDigest
    type: string
    value: sha256:1a45f2c9a21d8eb4046700c01b6e744d4e898ff2dd2c3eba871678880ef0fb88
  - name: unsignedWindowsDigest
    type: string
    value: sha256:925a6bf1bdc8d87c1b914d8dc91ac066178799098a95abbcc45f9b2565c7cd86
  - name: signedMacDigest
    type: string
    value: sha256:5d0b769571c52279cf937fa1f8c9db65b0c08b6c18fdbc16fcb0f8bacc2e606f
  - name: signedWindowsDigest
    type: string
    value: sha256:61b4b0bf553852ccbfb58a21d91461fe8a5206b2406719524a56a470101a9a7f
  - name: binaries_path
    type: string
    value: /workspace/data/e492dd56-18dc-4144-b83c-301c9b10d017/binaries/signed
```

File: tasks/sign-binaries/sign-binaries.yaml (hunk 5 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: there was another MR inroute by swickers earlier so I updated as such,  now I updated it as 3.0.0 as new steps and new result was added. please advise if that should be changed in any way

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: `CONTENT_DIR` path updated as such.

> There will be another task that will fetch the binaries from somewhere? 

yes extract-binaries-from-image task is expected to put binaries in that path. 

>  currently the workspace uses a PVC that is 1 GB in size. Will that be enough?

For now it should be enough but when bigger products such as OpenShift migrate to use this workflow we will need more storage. IIRC, each binary of that product is 1GB+ and there will be several binaries here for mac/windows and multiple architectures

File: tasks/sign-binaries/sign-binaries.yaml (hunk 8 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: leftover code from before, updated

File: tasks/sign-binaries/sign-binaries.yaml (hunk 5 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: OK. It's still work in progress, so probably no need to grow the versions this much before even the first working version. But I don't mind.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: OK, I guess we can worry about that later. I thought we had an option to provide a PV via data file somewhere, but I can't find it, so maybe not.

I wonder, couldn't the registry be also used for this? The extract-binaries-from-image would push the unsigned binaries to registry via oras. Or the extraction of the binaries could even be part of this task.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 8 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: ok. but the mocks should also use the same shebang as the actual task scripts.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: > The extract-binaries-from-image would push the unsigned binaries to registry via oras. Or the extraction of the binaries could even be part of this task.

I'm not following what that would gain us here?   
No matter what we'd need the binaries in the workspace at some point so we can generate checksums and push them to CDN.  I guess if CDN allowed a registry path as the source path then this could help us not need a larger PV.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 8 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: ack, shebang updated for tests.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 9 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: removed the todo line now, not needed.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: > I'm not following what that would gain us here?

We wouldn't need to worry about the size of the PV. Because you could use emptyDir which I think is only limited by disk space on the kubernetes node where the pod runs.

File: tasks/sign-binaries/tests/mac_mocks.sh [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+    scp_args=$@
+
+    # Check for Mac signing script
+    if [[ "$scp_args" == *"mac_signing_script"* ]]; then
+        echo "MAC SCP command detected with args: $@"
+        echo "$@" > $(workspaces.data.path)/scp_mac_args.txt
+    fi
+
+    # Check for Windows signing script
```
Comment (mmalina) [inline tasks/sign-binaries/tests/mac_mocks.sh]: I'm not sure exactly how the various mock files differ, but I can see that this one is only used for the mac signing step, so this shouldn't be here, right? And then `mocks.sh` is not used for mac signing, so it shouldn't have the mac signing script condition.

Comment (pkhander) [reply tasks/sign-binaries/tests/mac_mocks.sh]: Main reason is to keep them separate, since we were working on separate step implementation that would cause less conflicts.
You are correct about extra windows and mac tests, I updated that in the latest commit, ty!

File: tasks/publish-to-cgw/publish-to-cgw.yaml
Comment (mmalina) [inline tasks/publish-to-cgw/publish-to-cgw.yaml:1]: Can we move the python script to [the utils repo](https://github.com/konflux-ci/release-service-utils)? I didn't realize this was here until now. In the utils repo, you can add proper unit tests and we also have black and linting set up. If you want to do it later, can you please create a ticket for that?

Comment (pkhander) [reply tasks/publish-to-cgw/publish-to-cgw.yaml:1]: yea sure! it was already moved to the utils repo as [developer-portal-wrapper](https://github.com/konflux-ci/release-service-utils/tree/main/developer-portal-wrapper) by @scoheb 

I think he had made some modifications for it to work for RHEL-AI using internal-requests, I will make sure it is backwards compatible with this task as well. I have created a jira ([CLDX-182](https://issues.redhat.com/browse/CLDX-182)) to make sure we use that version for this task. I would like to get this in if possible and test separately if all works good with using that version.

Comment (mmalina) [reply tasks/publish-to-cgw/publish-to-cgw.yaml:1]: OK, sounds good, thanks.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: > CONTENT_DIR path updated as such.

It is correct in the signing task, but now it's incorrect in tasks/publish-to-cgw/publish-to-cgw.yaml:
```
CONTENT_DIR = "$(params.contentDir)"
```
The parameter should always be relative.

Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: That was updated in publish-to-cgw.yaml because of how sign-binares returned it. 

I have updated both to match what you have said.

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: Now it's ok, thanks.

**Review by mmalina — DISMISSED** [review]

File: tasks/sign-binaries/sign-binaries.yaml [hunk 16]
```diff
@@ -71,44 +71,239 @@ spec:
...
+        set -eux
+
+        CONTENT_DIR=$(workspaces.data.path)/$(params.contentDir)
+        UNSIGNED_DIR=$CONTENT_DIR/unsigned
+        MAC_CONTENT=$UNSIGNED_DIR/macos
+        WINDOWS_CONTENT=$UNSIGNED_DIR/windows
+        LINUX_CONTENT=$CONTENT_DIR/linux
+
+        mkdir -p "$MAC_CONTENT" "$WINDOWS_CONTENT" "$LINUX_CONTENT"
+        cd "$CONTENT_DIR" || exit 1
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: The exit 1 is to make shellcheck happy, right? I think I saw that elsewhere.. @mmalina didn't we find some other way to make it pass?

File: tasks/sign-binaries/README.md
Comment (johnbieren) [inline tasks/sign-binaries/README.md:1]: If there is a slack channel or something y'all hang out on, it would be great to add it to the README (above the parameters) so we know where to go if a user has issues with this task and we (release service team) can't figure it out

File: tasks/sign-binaries/tests/pre-apply-task-hook.sh [hunk 17]
```diff
@@ -4,8 +4,9 @@
 TASK_PATH="$1"
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 yq -i '.spec.steps[0].script = load_str("'$SCRIPT_DIR'/mocks.sh") + .spec.steps[0].script' "$TASK_PATH"
-yq -i '.spec.steps[1].script = load_str("'$SCRIPT_DIR'/mocks.sh") + .spec.steps[1].script' "$TASK_PATH"
-yq -i '.spec.steps[2].script = load_str("'$SCRIPT_DIR'/checksum_mocks.sh") + .spec.steps[2].script' "$TASK_PATH"
+yq -i '.spec.steps[1].script = load_str("'$SCRIPT_DIR'/mac_mocks.sh") + .spec.steps[1].script' "$TASK_PATH"
+yq -i '.spec.steps[2].script = load_str("'$SCRIPT_DIR'/mocks.sh") + .spec.steps[2].script' "$TASK_PATH"
```
Comment (johnbieren) [inline tasks/sign-binaries/tests/pre-apply-task-hook.sh:8]: Cool stuff!

File: tasks/sign-binaries/sign-binaries.yaml [hunk 18]
```diff
@@ -71,44 +71,239 @@ spec:
...
 
-        # TODO CLDX-134
+        # Unzip files in each directory
+        for dir in unsigned/macos unsigned/windows linux; do
+          for file in "$dir"/*.gz; do
+            gunzip "$file"
+          done
+        done
+
+        cd "$UNSIGNED_DIR" || exit 1
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: Same comment as a few lines above, just opening a thread here too in case we have something better for it and don't forget this one

File: tasks/sign-binaries/README.md
Comment (pkhander) [reply tasks/sign-binaries/README.md:1]: Added!

File: tasks/sign-binaries/sign-binaries.yaml (hunk 16 above)
Comment (pkhander) [reply tasks/sign-binaries/sign-binaries.yaml]: Yeah right

File: tasks/sign-binaries/README.md
Comment (johnbieren) [reply tasks/sign-binaries/README.md:1]: Awesome, much appreciated

File: tasks/sign-binaries/sign-binaries.yaml (hunk 16 above)
Comment (johnbieren) [reply tasks/sign-binaries/sign-binaries.yaml]: Let's wait for Martin to respond tomorrow then I will either resolve or make the small adjustment and then I'll approve

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: Here's the shellcheck wiki page for this: https://www.shellcheck.net/wiki/SC2164
> Rationale:
> cd can fail for a variety of reasons: misspelled paths, missing directories, missing permissions, broken symlinks and more.

The issue is that what if somewhere later you assume you're in your dir, but you never moved there. And then you do `rm *` or something.

But all of this is only a concern if you allow the script to carry on when something fails. As always, I would recommend using `set -eux` or `set -eu` in the task unless you have a good reason you don't want it. With it, the shellcheck issue disappears.

**Review by johnbieren — DISMISSED** [review]

**Review by johnbieren — APPROVED** [review]

**Comment by johnbieren** [comment]
There is no e2e test for this yet. Merging
